slither/detectors/variables/possible_const_state_variables.py

```diff
@@ -0,0 +1,71 @@
+"""
+Module detecting state variables that could be declared as constant
+"""
+
+from slither.core.solidity_types.elementary_type import ElementaryType
+from slither.detectors.abstract_detector import AbstractDetector, DetectorClassification
+from slither.slithir.operations import OperationWithLValue
+
+
+class ConstCandidateStateVars(AbstractDetector):
+    """
+    State variables that could be declared as constant detector.
+    Not all types for constants are implemented in Solidity as of 0.4.25.
+    The only supported types are value types and strings (ElementaryType).
+    Reference: https://solidity.readthedocs.io/en/latest/contracts.html#constant-state-variables
+    """
+
+    ARGUMENT = 'const-candidates-state'
+    HELP = 'detect state variables that could be const'
+    IMPACT = DetectorClassification.INFORMATIONAL
+    CONFIDENCE = DetectorClassification.HIGH
+
+    @staticmethod
+    def lvalues_of_operations_with_lvalue(contract):
+        ret = []
+        for f in contract.functions:
+            if f.name == 'fallback':  # Skip fallback function to avoid false positives
+                continue
+            for n in f.nodes:
+                for ir in n.irs:
+                    if isinstance(ir, OperationWithLValue):
+                        ret.append(ir.lvalue)
+        return ret
+
+    @staticmethod
+    def non_const_state_variables(contract):
+        return [variable for variable in contract.state_variables if not variable.is_constant]
+
+    def detect_const_candidates(self, contract):
+        const_candidates = []
+        non_const_state_vars = self.non_const_state_variables(contract)
+        lvalues_of_operations = self.lvalues_of_operations_with_lvalue(contract)
+        for non_const in non_const_state_vars:
+            if non_const not in lvalues_of_operations \
+                    and non_const not in const_candidates \
+                    and isinstance(non_const.type, ElementaryType):
+                const_candidates.append(non_const)
+
+        return const_candidates
+
+    def detect(self):
+        """ Detect state variables that could be const
+        """
+        results = []
+        for c in self.contracts:
```

Comment:
Thank you for the hint. I will try this out asap.

Reply:
@montyly just to double check if I understood correctly as I want to make it right.

As you mentioned above, the variable should not be declared constant in the base class when we want to modify it in the derived, but what if we want to have it constant state both in base and derived? 

To clarify, when the detector iterates over contracts_derived, it does not take into account base contract and does not recognize base contracts' constant state candidates. I.e:

self.contracts gives us

```
$ slither tests/const_state_variables.sol --detect-const-candidates-state
INFO:Detectors: State variables that could be const in tests/const_state_variables.sol, Contract: A, Vars ['myFriendsAddress', 'test', 'text2']
INFO:Detectors: State variables that could be const in tests/const_state_variables.sol, Contract: B, Vars ['myFriendsAddress', 'test', 'text2', 'mySistersAddress']
INFO:Slither:tests/const_state_variables.sol analyzed (2 contracts), 2 result(s) found
```

slither.contracts_derived gives:

```
slither tests/const_state_variables.sol --detect-const-candidates-state
INFO:Detectors: State variables that could be const in tests/const_state_variables.sol, Contract: B, Vars ['myFriendsAddress', 'test', 'text2', 'mySistersAddress']
INFO:Slither:tests/const_state_variables.sol analyzed (2 contracts), 1 result(s) found
``` 

Could please have a look at tests/const_state_variables.sol and confirm if only contracts_derived are to be used? or is it better to detect candidates in both contracts to detect cases when the constant should be set in the base class?

Reply:
Oh good point.
So a `stateVariable` has a `.contract` attribute which returns the instance of the contract where its declared.

A solution could be to iterate over `slither.contracts_derived`, but to use `variable.contract.name` for the result, so that at the end you can group them per original contract, and have something like:
```
INFO:Detectors: State variables that could be const in tests/const_state_variables.sol, Contract: A, Vars ['myFriendsAddress', 'test', 'text2']
INFO:Detectors: State variables that could be const in tests/const_state_variables.sol, Contract: B, Vars ['mySistersAddress']
```